statsmodels/duration/_kernel_estimates.py

```diff
@@ -0,0 +1,204 @@
+
+        # Dimension reduction step
+        if dimred:
+            sfe = PHReg(time, exog, status0).fit()
+            fitval_e = sfe.predict().predicted_values
+            sfc = PHReg(time, exog, 1 - status0).fit()
+            fitval_c = sfc.predict().predicted_values
+            exog2d = np.hstack((fitval_e[:, None], fitval_c[:, None]))
+            exog2d -= exog2d.mean(0)
+            exog2d /= exog2d.std(0)
+        else:
+            exog2d = exog
+
+        ip0 = 0
+        for i in range(nobs):
+
+            if k == 0:
+                kd1 = exog2d - exog2d[i, :]
+                kd1 = kfunc(kd1)
+                kd[i] = kd1
+
+            # Get the local all-causes survival function
+            if k == 0:
+                denom = np.cumsum(kd[i][::-1])[::-1]
+                num = kd[i] * statusa
+                rat = num / denom
+                tr = 1e-15
+                ii = np.flatnonzero((denom < tr) & (num < tr))
+                rat[ii] = 0
+                ratc = 1 - rat
+                ratc = np.clip(ratc, 1e-10, np.inf)
+                lrat = np.log(ratc)
+                prat = np.cumsum(lrat)[ie]
+                sf = np.exp(prat)
+                sp[i] = np.r_[1, sf[:-1]]
+                n_risk[i] = denom[ie]
+
+            # Number of cause-specific deaths at each unique time.
+            d0 = np.bincount(rtime, weights=status0*kd[i],
+                             minlength=len(utime))
+
+            # The cumulative incidence function probabilities.  Carry
+            # forward once the effective sample size drops below 1.
+            ip1 = np.cumsum(sp[i] * d0 / n_risk[i])
+            jj = len(ip1) - np.searchsorted(n_risk[i][::-1], 1)
+            if jj < len(ip1):
+                ip1[jj:] = ip1[jj - 1]
+            if freq_weights is None:
+                ip0 += ip1
+            else:
+                ip0 += freq_weights[i] * ip1
+
+        if freq_weights is None:
+            ip0 /= nobs
+
+        ip.append(ip0)
+
+    return utime, ip
+
+
+def _kernel_survfunc(time, status, exog, kfunc, freq_weights):
+    """
+    Estimate the marginal survival function under dependent censoring.
+
+    Parameters
+    ----------
+    time : array-like
+        The observed times for each subject
+    status : array-like
+        The status for each subject (1 indicates event, 0 indicates
+        censoring)
+    exog : array-like
+        Covariates such that censoring is independnet conditioned on
+        exog
+    kfunc : function
+        Kernel function
+    freq_weights : array-like
+        Optional frequency weights
+
+    Returns
+    -------
+    probs : array-like
+        The estimated survival probabilities
+    times : array-like
+        The times at which the survival probabilities are estimated
+
+    References
+    ----------
+    Estimating marginal survival function by adjusting for dependent
+    censoring using many covariates
+    http://arxiv.org/pdf/math/0409180.pdf
```

Comment:
full citation

Zeng, Donglin. 2004. “Estimating Marginal Survival Function by Adjusting for Dependent Censoring Using Many Covariates.” The Annals of Statistics 32 (4): 1533–55. doi:10.1214/009053604000000508.